### PR TITLE
Haptic and solenoid cleanup

### PR DESCRIPTION
**Pull request by purdeaandrei**
## Description

Overall this PR fixes:
* how haptic configuration is first set up if the eeprom is empty/corrupt, or if previous firmware didn't have haptic enabled.
* a bunch of small bugs related to solenoid (reset, loading of settings from eeprom, etc...)
* makes solenoid a little more configurable

For more detailed descriptions please see my individual commits for descriptions of my changes.

## Types of Changes

- [x] Core
- [x] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

N/A (i think)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by purdeaandrei** [comment]
@zvecr I pushed a fix for the build issue.
I would like to argue that the behavior change is a bugfix, and it never should have allowed the dwell time to overflow, and get a less-then-MIN value.
Before this dwell time wasn't really saved in eeprom while it was clear from code that that was the intent. Would the fixing of that also make this PR a breaking change?

**Comment by purdeaandrei** [comment]
Also, there are no keyboards within the QMK tree, that use the core solenoid functionality.
(There's some mtdjr user code for solenoid but that is different, looks to be the originator of QMK solenoid code)
The functionality of saving/restoring to eeprom, and resetting the solenoid settings was so broken, that I find it hard
to believe anyone's using it out of tree in its current form.

**Comment by zvecr** [comment]
We care about out of tree breakages too, some of the docs might need a little update here and there but it's mostly covered by <https://docs.qmk.fm/#/breaking_changes_instructions>. I will leave it for other collabs to express their stance. However i would add that an outside view of issue classification and chosen workflow might not align with the direction of the project.

File: docs/feature_haptic_feedback.md
Comment (drashna) [inline docs/feature_haptic_feedback.md:52]: I'm not sure if I like the name for this, but I'm not sure if there is a better way to name this.

Comment (purdeaandrei) [reply docs/feature_haptic_feedback.md:52]: Well BUZZ is used in other places such as the HPT_BUZ keycode "Toggle solenoid buzz on/off".
If we were to rename this, then we would need to rename it anywhere.
If Buzz is not a good name, than maybe rattle?
keycode would become HPT_RAT. macro would be SOLENOID_DEFAULT_RATTLE?

**Comment by drashna** [comment]
>Also, there are no keyboards within the QMK tree, that use the core solenoid functionality.
That's not actually true. I have several keymaps that have solenoids and use the feature.

However, I'm not sure what the buz part is supposed to do. I don't really notice a change in behavior.

**Comment by purdeaandrei** [comment]
@drashna 

I probably made a mistake while initially looking for keyboards/keymaps using the solenoid code, there are a few using it in the QMK tree.

> I'm not sure what the buzz part is supposed to do. I don't really notice a change in behavior.

I did not write this code initially, but based on reading the code, I believe it means that in buzz mode the solenoid is tapped multiple times for each key press.

In mtdjr's non-core code in users/mtdjr/solenoid.h SOLENOID_MIN_DWELL was only used to specify half the period for the buzzing solenoid signal.
So the macro meanings were:
* SOLENOID_DEFAULT_DWELL - the starting dwell time setting. (in the buzz case this specifies the length of the entire waveform)
* SOLENOID_MAX_DWELL - this specified the maximum dwell time, beyond which the solenoid dwell time was not allowed to be incremented when adjusting it
* SOLENOID_MIN_DWELL - this is NOT the opposite of SOLENOID_MAX_DWELL in the (what appears to be) original mtdjr code.

I think someone took the mtdjr code and driverized it (imported into core), and during driverization, some things changed:
* using eeprom to store settings (incorrectly implemented)
* SOLENOID_MIN_DWELL still means the same as it used to mean in the original code, but now it also means a lower limit when adjusting the solenoid dwell time. In the original code the lower limit was just zero.
* the adjustment code was changed to go round robin, although incorrectly.
* (not complete list)

So to see buzz do anything special, for example you need to:
0) Use the code on this branch. There's lots of bugs otherwise.
1) Have buzz enabled by using the HPT_BUZ key
2) increase solenoid dwell time to maybe 8 times the SOLENOID_MIN_DWELL_TIME setting. This will make the solenoid tap 4 times for each keypress. I don't know if you will hear the difference, but I tested this with a logic analyzer and it works like this.

Again, without this PR there are plenty of bugs. For instance the reset doesn't reset. The saved settings (such as solenoid dwell time) aren't properly reloaded on power cycle. I noted that by default solenoid buzz was on,.so you won't actually see anything increasing dwell time, except if you increase it beyond 2 times the minimum. Please see my individual commit messages for more detailed descriptions of the bugs that I fixed.

Andrei

**Comment by purdeaandrei** [comment]
@drashna Alright, so since this change is still categorized as a breaking change, it will definitely have to be retargeted to develop. So I'm gonna do that now...

**Comment by purdeaandrei** [comment]
Alright, I rebased it to develop, please let me know if there are any more pending actions on me for this PR.

**Comment by purdeaandrei** [comment]
@noroadsleft What happened here? The force-push screwed up my PRs history... 
Is that force-push correct, or a mistake?
Do I need to rebase again?

**Comment by purdeaandrei** [comment]
@noroadsleft  I see commit 0f9b7b9768e53add15dc7e860efab0464cd7533d was inserted underneath a stack of existing stack of commits that were already on develop, so that's why the force push was done. Any reason why that commit itself could not have been rebased, to not edit history?

**Comment by drashna** [comment]
IIRC, a "simple" way to fix this is with:
```
git reset upstream/develop --hard
git cherrypick d2857d3^...739cd61
git push --force-with-lease
```
(the "with lease" part isn't strictly necessary, it's just good habit, IMO)

You can test this with another branch, and see if that cleans up things, first.  And then do to this branch.

**Comment by noroadsleft** [comment]
> I see commit [0f9b7b9](https://github.com/qmk/qmk_firmware/commit/0f9b7b9768e53add15dc7e860efab0464cd7533d) was inserted underneath an existing stack of commits that were already on develop, so that's why the force push was done. Any reason why that commit itself could not have been rebased, to not edit history?

QMK's workflow for the `develop` branch is that pull requests that target the branch are merged the same way PRs to `master` are, and then the `develop` branch is rebased upon `master` once a week. Unfortunately the result is that long-standing PRs to `develop` also need to be rebased regularly.

You can follow the instructions drashna gave above in situations where you need to rebase because the target branch (`develop`) has changed.

We usually try to rebase `develop` every Friday at 12:00 US Pacific Time (UTC -7:00 at the moment).

Speaking of which, it needs to be rebased again...

**Comment by purdeaandrei** [comment]
Rebased

**Review by noroadsleft — APPROVED** [review]
Slight `__attribute__ ((weak))` :heavy_check_mark: because I don't have any experience with the Haptic or Solenoid features, but the code delta looks good to me.

**Comment by purdeaandrei** [comment]
Is there anything I can do to help this progress forward?

**Comment by purdeaandrei-movi** [comment]
Hello,

Would it help if I asked some of my firmware's users who use a solenoid to comment here about my solenoid fixes?

I understand they are not official approvers, but maybe it will increase approver confidence in the changeset?

Or is there anything else I can do to make this move forward? It's been almost 3 months since the last commit (there were only rebases after that, the last rebase almost a month ago).

Andrei

**Comment by purdeaandrei** [comment]
Hello,

Would it help if I asked some of my firmware's users who use a solenoid to comment here about my solenoid fixes?

I understand they are not official approvers, but maybe it will increase approver confidence in the changeset?

Or is there anything else I can do to make this move forward? It's been almost 3 months since the last commit (there were only rebases after that, the last rebase almost a month ago).

Andrei

**Comment by drashna** [comment]
I think it should be okay (I have tested on a couple of my boards, and it seems to be fine).

**Comment by noroadsleft** [comment]
@drashna I'd be okay with merging this if it has your approval. I can't test this at all (don't own any haptic or solenoid devices).

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**
